pkg/zfs/volume.go

```diff
@@ -117,22 +147,67 @@ func UpdateZvolInfo(vol *apis.ZFSVolume) error {
 		return nil
 	}
 
-	newVol, err := builder.BuildFrom(vol).
+	newVol, err := volbuilder.BuildFrom(vol).
 		WithFinalizer(finalizers).
 		WithLabels(labels).Build()
 
 	if err != nil {
 		return err
 	}
 
-	_, err = builder.NewKubeclient().WithNamespace(OpenEBSNamespace).Update(newVol)
+	_, err = volbuilder.NewKubeclient().WithNamespace(OpenEBSNamespace).Update(newVol)
 	return err
 }
 
 // RemoveZvolFinalizer adds finalizer to ZFSVolume CR
 func RemoveZvolFinalizer(vol *apis.ZFSVolume) error {
 	vol.Finalizers = nil
 
-	_, err := builder.NewKubeclient().WithNamespace(OpenEBSNamespace).Update(vol)
+	_, err := volbuilder.NewKubeclient().WithNamespace(OpenEBSNamespace).Update(vol)
+	return err
+}
+
+// GetZFSSnapshotStatus returns ZFSSnapshot status
+func GetZFSSnapshotStatus(snapID string) (string, error) {
+	getOptions := metav1.GetOptions{}
+	snap, err := snapbuilder.NewKubeclient().
+		WithNamespace(OpenEBSNamespace).Get(snapID, getOptions)
+
+	if err != nil {
+		return "", err
+	}
+
+	return snap.Spec.Status, nil
+}
+
+// UpdateSnapInfo updates ZFSSnapshot CR with node id and finalizer
+func UpdateSnapInfo(snap *apis.ZFSSnapshot) error {
+	finalizers := []string{ZFSFinalizer}
+	labels := map[string]string{ZFSNodeKey: NodeID}
+
+	if snap.Finalizers != nil {
+		return nil
+	}
+
+	newSnap, err := snapbuilder.BuildFrom(snap).
+		WithFinalizer(finalizers).
+		WithLabels(labels).Build()
+
+	// set the status to ready
+	newSnap.Spec.Status = ZFSStatusReady
+
+	if err != nil {
+		return err
+	}
+
+	_, err = snapbuilder.NewKubeclient().WithNamespace(OpenEBSNamespace).Update(newSnap)
+	return err
+}
+
+// RemoveSnapFinalizer adds finalizer to ZFSSnapshot CR
+func RemoveSnapFinalizer(snap *apis.ZFSSnapshot) error {
+	snap.Finalizers = nil
+
+	_, err := snapbuilder.NewKubeclient().WithNamespace(OpenEBSNamespace).Update(snap)
```

Comment:
We should be removing only the finalizers which we are supposed to. 
Think of a case where some client has put its own implementation/automation and put their own finalizer
to carry out other bookkeeping!

Reply:
good, point. Right now there is only one owner which is node agent and there can not be anyone else owning that, hence setting it to null when destrying.